### PR TITLE
[test] Add MPI halo cell exchange benchmark

### PR DESCRIPTION
**Pull request by ajocksch**

### CODE REVIEW

**Comment by vkarak** [comment]
@ajocksch Why is this still a WIP?

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    if (ierr!=0) exit(1);
+    end = 0;
+    while (end == 0){
+        if (comm_rank==0){
+            end = (fgets(inputbuf, sizeof(inputbuf), stdin) == NULL);
+            if (end == 0){
+                pinputbuf = inputbuf;
+                while (pinputbuf[0]==' ') pinputbuf++;
+                end = !((pinputbuf[0]>='0')&&(pinputbuf[0]<='9'));
+                pinputbuf = inputbuf;
```
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Why don't you simply do an `fscanf()` here?

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
+#include <stdlib.h>
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Please add a small description as of what this benchmark is doing exactly. Also, add comments in the code explaining briefly the different phases.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                        if (ierr!=0) exit(1);
+                        ierr = MPI_Isend(sendbuf+(i*2+1)*halosize[i]*sizeof(char), halosize[i], MPI_CHAR, rank_dest, 1, cart_comm, request+i*2+1+ndims*2);
+                        if (ierr!=0) exit(1);
+                    }
+                    ierr = MPI_Waitall(ndims*2*2, request, status);
+                    if (ierr!=0) exit(1);
+                }
+
+                stop = MPI_Wtime ();
+                ttt = stop - start;
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Don't be so cryptical in variable names:

```suggestion
                elapsed_time = stop - start;
```

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                free(recvbuf);
+                free(sendbuf);
+                ierr = MPI_Comm_free(&cart_comm);
+                if (ierr!=0) exit(1);
+                ierr = MPI_Comm_free(&red_comm);
+                if (ierr!=0) exit(1);
+            }
+        }
+    }
+    ierr = MPI_Finalize();
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: `ierr` is not used here.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    char *sendbuf, *recvbuf, inputbuf[1000], *pinputbuf;
+    double start, stop, deltatmin, deltatmax, ttt;
+    int rank_source, rank_dest, i, j;
+
+    if (argc>1){
+        printf("%s < ndims dim1 dim2 ... halosize1 halosize2 ...\n", argv[0]);
+        exit(0);
+    }
+    ierr = MPI_Init(NULL, NULL);
+    if (ierr!=0) exit(1);
```
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Also do not use `ierr` as name. It is really Fortran nomenclature and the `i` comes from a coding style decision about specifying the type of the variable in the name, which is a bad idea. Simply use `err`.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    ierr = MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank);
+    if (ierr!=0) exit(1);
+    end = 0;
+    while (end == 0){
+        if (comm_rank==0){
+            end = (fgets(inputbuf, sizeof(inputbuf), stdin) == NULL);
+            if (end == 0){
+                pinputbuf = inputbuf;
+                while (pinputbuf[0]==' ') pinputbuf++;
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]:
```suggestion
                while (*pinputbuff==' ') pinputbuf++;
```

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                end = !((pinputbuf[0]>='0')&&(pinputbuf[0]<='9'));
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Similarly here.

**Comment by ajocksch** [comment]
the code also compiles with C++ compilers; when reading a line of the input the number of columns is not known in advance

File: cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        self.num_gpus_per_node = 0
+
+        self.build_system.cflags = ['-O2']
+
+        self.executable_opts = ['< input.txt']
+
+        self.sanity_patterns = sn.all([
+            sn.assert_eq(
+                sn.count(sn.findall(r'halo_cell_exchange',
+                                    self.stdout)), 9)
```
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py]: I don't see an explanation for the `9` here? Is it fixed or related to the `num_tasks`?

File: cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                'time_2_10000': (9.721279e-02, None, 0.50, 's'),
+                'time_2_1000000': (4.934530e+00, None, 0.50, 's'),
+                'time_4_10': (5.878997e-02, None, 0.50, 's'),
+                'time_4_10000': (1.495080e-01, None, 0.50, 's'),
+                'time_4_1000000': (6.791397e+00, None, 0.50, 's'),
+                'time_6_10': (5.428815e-02, None, 0.50, 's'),
+                'time_6_10000': (1.540580e-01, None, 0.50, 's'),
+                'time_6_1000000': (9.179296e+00, None, 0.50, 's')
+            },
+            'kesch:cn': {
```
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py]: This is not addressed. You may simply remove `kesch` from supported systems if you don't want to put numbers here.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 10]
```diff
@@ -1,156 +1,332 @@
...
-    int rank_source, rank_dest, i, j;
+#define NCALLS 10000
 
-    if (argc>1){
-        printf("%s < ndims dim1 dim2 ... halosize1 halosize2 ...\n", argv[0]);
-        exit(0);
+int
+main (int argc, const char *argv[])
+{
+  MPI_Comm cart_comm, red_comm;
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: Which formatting style did you use? One-space identation is... not an identation actually.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 11]
```diff
@@ -1,156 +1,332 @@
...
+    }
+  if (comm_rank == 0)
+    {
+      if (strcmp (argv[1], "-") == 0)
+	{
+	  pFile = stdin;
+	}
+      else
+	{
+	  pFile = fopen (argv[1], "r");
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]: You should normally check here that `pFile` is not `NULL`.

File: cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c [hunk 12]
```diff
@@ -1,156 +1,332 @@
...
+	}
+    }
+  end = 0;
+  while (end == 0)
```
Comment (vkarak) [inline cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c]:
```c
while (!end)
```
is more readable here I believe.

File: cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ajocksch) [reply cscs-checks/microbenchmarks/mpi/halo_cell_exchange.py]: it is the number of lines (subtests) of the input file

**Pull request merged**